### PR TITLE
Added `wait` api

### PR DESCRIPTION
**Pull request by delvedor**
As titled, in this way we can provide a better support for *async-await*.
See https://github.com/fastify/fastify/issues/481 for context.

### CODE REVIEW

**Comment by StarpTech** [comment]
Nice :ok_hand:

File: test/basic.test.js [hunk 1]
```diff
@@ -205,3 +205,47 @@ test('promise long resolve', (t) => {
     t.notOk(err)
   })
 })
+
+test('wait plugin registration', (t) => {
```
Comment (StarpTech) [inline test/basic.test.js]: What about test to cover error propagations with async await ?

```js
app.use(async () => {})
```

Comment (delvedor) [reply test/basic.test.js]: Take a look [here](https://github.com/mcollina/avvio/blob/74ac2125913658002a21169bae690f417e0cfe98/test/async-await.js#L145-L158) :D

Comment (StarpTech) [reply test/basic.test.js]: I didn't saw it :D

File: boot.js [hunk 2]
```diff
@@ -206,6 +206,23 @@ Boot.prototype.after = function (func, cb) {
   return this
 }
 
+Boot.prototype.wait = function (func) {
```
Comment (StarpTech) [inline boot.js]: `func` can be removed

File: README.md [hunk 3]
```diff
@@ -261,10 +262,23 @@ app.ready(function (err, context, done) {
 
 Returns the instance on which `ready` is called, to support a chainable API.
 
+-------------------------------------------------------
+<a name="wait"></a>
+
+### app.wait()
+If you are using async await, you can avoid the use of `after` and `ready`, and just use the `wait` api.
```
Comment (StarpTech) [inline README.md]: If you are using async await, you can avoid the use of `after` and `ready`, and just use the `wait` api which returns a promise.

**Comment by StarpTech** [comment]
LGTM

**Comment by mcollina** [comment]
Closing for now, feel free to reopen if you still want to pursue this idea.

**Comment by allevo** [comment]
Can we reopen this discussion?

From developer point of view, this feature is very useful. For example:

```
module.exports = async function (fastify, options) {
  fastify.register(fastifyEnv, { ... })
  await fastify.wait()
  fastify.register(fastifyMongodb, { url: fastify.config.MONGODB_URL })
}
```

currently writable as:

```
module.exports = function (fastify, options, done) {
  fastify.register(fastifyEnv, { ... })
    .after(() => {
      fastify.register(fastifyMongodb, { url: fastify.config.MONGODB_URL })
        .after((err, done) => {
          if (err) return done(err)
          fastify.mongo.db.createCollection('myCollection', done)
        })
    })
  done()
}
```

**Comment by delvedor** [comment]
The only way to support what you are proposing is to play a lot with the internal queues, which is not that easy and it will require a crazy amount of tests.

**Comment by allevo** [comment]
@delvedor I know.

But the process start up is not possible without doing a callback hell...

**Comment by mcollina** [comment]
How about?

```js
module.exports = async function (fastify, options) {
  fastify
     .register(fastifyEnv, { ... })
     .register(fastifyMongodb, { url: fastify.config.MONGODB_URL })
     .register(myRoutes)
}

async function myRoutes(f, opts) {
  fastify.mongo.db.createCollection('myCollection', done)
}
```

**Comment by mcollina** [comment]
My bad, that's not possible.

**Comment by mcollina** [comment]
or better:

```js
fastify
  .register(fastifyEnv)
  .register(addMongo)
  .register(myRoutes)

async function addMongo (f, opts) {
     f.register(fastifyMongodb, { url: fastify.config.MONGODB_URL })
}

async function myRoutes(f, opts) {
  fastify.mongo.db.createCollection('myCollection', done)
}
```

**Comment by allevo** [comment]
This example doesn't work
```

async function decorateCollectionInstance (fastify) {
  fastify.decorate('collection', new MyAwesomeBusinessLogicClass(fastify.mongo.db)
  // TypeError: Cannot read property 'db' of undefined
}

async function mongodb (fastify) {
  fastify.register(fastifyMongodb, { url: fastify.config.MONGODB_URL })
  fastify.register(decorateCollectionInstance)
}

const envSchema = {
  type: 'object',
  required: ['MONGODB_URL'],
  properties: {
    MONGODB_URL: { type: 'string' }
  }
}
module.exports = async function (fastify, options) {
  fastify.register(fastifyEnv, { schema: envSchema, data: options, env: false })
    .register(mongodb)
    .register(decorateCollectionInstance)
}

```

**Comment by mcollina** [comment]
@allevo can you send a PR for fastify with that example? That _should_ work.

**Pull request closed without merging**